mixnet/src/fragment.rs

```diff
@@ -1,22 +1,249 @@
+                    },
+                    body: Vec::from(chunk),
+                })
+                .collect(),
+        ))
+    }
+
+    fn num_chunks(msg: &[u8], chunk_size: usize) -> usize {
+        msg.len() / chunk_size + (msg.len() % chunk_size > 0) as usize
+    }
+}
+
+impl Deref for FragmentSet {
+    type Target = Vec<Fragment>;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
 #[derive(PartialEq, Eq, Debug, Clone)]
-pub(crate) struct Fragment {}
+pub(crate) struct Fragment {
+    header: FragmentHeader,
+    body: Vec<u8>,
+}
 
 impl Fragment {
+    pub(crate) fn bytes(&self) -> Vec<u8> {
+        let mut out = Vec::with_capacity(FRAGMENT_HEADER_SIZE + self.body.len());
+        out.extend(self.header.bytes());
+        out.extend(&self.body);
+        out
+    }
+
     pub(crate) fn from_bytes(value: &[u8]) -> Result<Self, MixnetError> {
-        todo!()
+        Ok(Self {
+            header: FragmentHeader::from_bytes(&value[0..FRAGMENT_HEADER_SIZE])?,
+            body: value[FRAGMENT_HEADER_SIZE..].to_vec(),
+        })
     }
 }
 
-pub struct MessageReconstructor {}
+type FragmentSetId = Uuid;
+type FragmentId = u8;
+const MAX_NUM_FRAGMENTS: usize = FragmentId::MAX as usize + 1;
+const FRAGMENT_HEADER_SIZE: usize = 18;
+
+#[derive(PartialEq, Eq, Debug, Clone)]
+struct FragmentHeader {
+    set_id: FragmentSetId,
+    total_fragments: FragmentId,
+    fragment_id: FragmentId,
+}
+
+impl FragmentHeader {
+    fn bytes(&self) -> [u8; FRAGMENT_HEADER_SIZE] {
+        let mut out = [0u8; FRAGMENT_HEADER_SIZE];
+        out[0..16].copy_from_slice(self.set_id.as_bytes());
+        out[16] = self.total_fragments;
+        out[17] = self.fragment_id;
+        out
+    }
+
+    fn from_bytes(value: &[u8]) -> Result<Self, MixnetError> {
+        if value.len() != FRAGMENT_HEADER_SIZE {
+            return Err(MixnetError::InvalidFragmentHeader);
+        }
+
+        Ok(Self {
+            set_id: Uuid::from_slice(&value[0..16])?,
+            total_fragments: value[16],
+            fragment_id: value[17],
+        })
+    }
+}
+
+pub struct MessageReconstructor {
+    fragment_sets: HashMap<FragmentSetId, FragmentSetReconstructor>,
+}
 
 impl MessageReconstructor {
     pub fn new() -> Self {
-        todo!()
+        Self {
+            fragment_sets: HashMap::new(),
+        }
     }
 
     pub fn add(&mut self, fragment: Fragment) -> Option<Vec<u8>> {
-        todo!()
+        let set_id = fragment.header.set_id;
+        self.fragment_sets
+            .entry(set_id)
+            .or_insert(FragmentSetReconstructor::new(
+                fragment.header.total_fragments,
+            ))
+            .add(fragment)
+            .map(|msg| {
+                // A message has been reconstructed completely from the fragment set.
+                // Delete the fragment set from the reconstructor.
+                self.fragment_sets.remove(&set_id);
+                msg
+            })
```

Comment:
Nitpick: This last `map` feels a bit weird, I think it would be better read as a `let Some(msg)` instead. As we are calling `self.fragment_sets` inside a call to a `self.fragment_sets`

Reply:
ba02c80 yeah, it can be confusing.